caffe/src/caffe/layers/molgrid_data_layer.cpp

```diff
@@ -1039,7 +1039,7 @@ void MolGridDataLayer<Dtype>::copyToBlobs(const vector<Blob<Dtype>*>& top, bool
 //set in memory buffer
 //will apply translate and rotate iff rotate is valid
 template <typename Dtype>
-void MolGridDataLayer<Dtype>::setReceptor(const vector<atom>& receptor, const vec& translate, const qt& rotate ) {
+void MolGridDataLayer<Dtype>::setReceptor(const vector<atom>& receptor, const vector<vec>& coords, const vec& translate, const qt& rotate ) {
```

Comment:
atom already has coordinates in it, what is the purpose of coords?  This could use a comment.

Reply:
The problem is that the coordinates in `atom` and in `model::coords` do not match. I haven't figured out why yet, but I went with the same API as `setLigand` (for which atoms are passed for types etc. but coordinates are passed explicitly from `model::coords`).

Reply:
perhaps we should change the api to take a types vector and a coordinates vector

Reply:
Yes, but we also need the `radii`

Reply:
i believe atom has the _starting_ coordinates

Reply:
When I was printing them out, many coordinates of the flexible residues where clustered far from the receptor, therefore I'm not 100% sure what was going on... This is why gradients were exactly zero. If they were starting coordinates they should have been inside of the box (and led to erroneous but non-zero gradients), I think.

![grad](https://user-images.githubusercontent.com/11348981/62385933-4b7bec80-b54e-11e9-9d21-f6ef5f9c9e14.png)

Reply:
the starting coordinates wouldn't have been inside the box if a random rigid body translation/rotation had been applied, and you were printing out intermediate coordinates with that transformation and comparing against initial coordinates prior to it

Reply:
True, but I was using `--cnn_freeze_receptor`. Are there rotations or translations not controlled by `--cnn_freeze_receptor`?

I can print out again the old coordinates and gradients and investigate this further.

Reply:
I suggest we change the `MolGridDataLater::setReceptor` and `MolGridDataLater::setLigand` as follows:

```
void MolGridDataLayer<Dtype>::setLigand(const vector<vec>& coords, const vector<float>& types, const vector<float>& radii, const bool calcCenter)

void MolGridDataLayer<Dtype>::setReceptor(const vector<vec>& coords, const vector<float>& types, const vector<float>& radii, const vec& translate, const qt& rotate )
```

The `CNNScorer::setReceptor` and `CNNScorer::setLigand` will then extract this information for the receptor and ligand, respectively, and pass it to the `MolGridDataLayer`.

@dkoes can/should I proceed with such changes?

Reply:
Yes, I think that is nice general api

Reply:
@dkoes After some experimentation, I finally decided for a different API:
```
setLigand(const vector<float3>& coords, const vector<smt>& smtypes, bool calcCenter)
```
i.e. the original `smina` types are passed. This avoids "pollution" to `cnn_scorer.h` and `test_utils.h` since to extract the numerical atom types and the radii they would need `std::shared_ptr<libmolgrid::AtomTyper> ligTypes` and the inclusion of some `libmolgrid` headers.

I hope this is ok, but let me know if it's not.

Reply:
(Same goes for `setReceptor`)

Reply:
Resolving because outdated (see changes).